File: IOS/DTOs/OptionSet/CredentialsSet.swift
/// Если вы правите это обязательно поправьте маппер в приложении.
public struct CredentialsSet: Codable, Hashable {
    public var clientCreds     : Set<ClientCredentials>
    public var salonCreds      : Set<SalonCredentials>
    public var employeeCreds   : Set<EmployeeCredentials>
    public var procedureCreds  : Set<ProcedureCredentials>
    public var positionCreds   : Set<PositionCredentials>
    public var worktimeCreds   : Set<WorktimeCredentials>
    public var appointmentCreds: Set<AppointmentCredentials>
    public var statisticCreds  : Set<StatisticCredentials>
    public var notifyCreds     : Set<NotificationCredentials>
    public var financeCreds    : Set<FinanceCredentials>
    public var salaryCreds     : Set<SalaryCredentials>
    
    public enum ClientCredentials: Codable {
        case edit
        case visibleBaseInfo
        case visibleContacts
    }

    public enum SalonCredentials: Codable {
        case editBaseInfo
    }

    public enum EmployeeCredentials: Codable {
        case edit
        case visible
    }

    public enum ProcedureCredentials: Codable {
        case ownEdit
        case otherEdit
    }

    public enum PositionCredentials: Codable {
        case edit
    }

    public enum WorktimeCredentials: Codable {
        case ownEdit
        case otherEdit
    }

    /// ----------------------

    public enum AppointmentCredentials: Codable {
       case all
    }

    public enum StatisticCredentials: Codable {
        case all
    }

    public enum NotificationCredentials: Codable {
        case all
    }

    public enum FinanceCredentials: Codable {
        case all
    }

    public enum SalaryCredentials: Codable {
        case all
    }
}

public extension CredentialsSet {
    static var all: CredentialsSet = .init(
        clientCreds     : [.edit, .visibleBaseInfo, .visibleContacts],
        salonCreds      : [.editBaseInfo],
        employeeCreds   : [.edit, .visible],
        procedureCreds  : [.otherEdit, .ownEdit],
        positionCreds   : [.edit],
        worktimeCreds   : [.otherEdit, .ownEdit],
        appointmentCreds: [.all],
        statisticCreds  : [.all],
        notifyCreds     : [.all],
        financeCreds    : [.all],
        salaryCreds     : [.all]
    )
    
    static var none: CredentialsSet = .init(
        clientCreds     : [],
        salonCreds      : [],
        employeeCreds   : [],
        procedureCreds  : [],
        positionCreds   : [],
        worktimeCreds   : [],
        appointmentCreds: [],
        statisticCreds  : [],
        notifyCreds     : [],
        financeCreds    : [],
        salaryCreds     : []
    )
}

public extension CredentialsSet {
    mutating func insert(_ c: ClientCredentials)       { self.clientCreds.insert(c) }
    mutating func insert(_ c: SalonCredentials)        { self.salonCreds.insert(c) }
    mutating func insert(_ c: EmployeeCredentials)     { self.employeeCreds.insert(c) }
    mutating func insert(_ c: ProcedureCredentials)    { self.procedureCreds.insert(c) }
    mutating func insert(_ c: PositionCredentials)     { self.positionCreds.insert(c) }
    mutating func insert(_ c: WorktimeCredentials)     { self.worktimeCreds.insert(c) }
    mutating func insert(_ c: AppointmentCredentials)  { self.appointmentCreds.insert(c) }
    mutating func insert(_ c: StatisticCredentials)    { self.statisticCreds.insert(c) }
    mutating func insert(_ c: NotificationCredentials) { self.notifyCreds.insert(c) }
    mutating func insert(_ c: SalaryCredentials)       { self.salaryCreds.insert(c) }
    mutating func insert(_ c: FinanceCredentials)      { self.financeCreds.insert(c) }
    
    mutating func remove(_ c: ClientCredentials)       { self.clientCreds.remove(c) }
    mutating func remove(_ c: SalonCredentials)        { self.salonCreds.remove(c) }
    mutating func remove(_ c: EmployeeCredentials)     { self.employeeCreds.remove(c) }
    mutating func remove(_ c: ProcedureCredentials)    { self.procedureCreds.remove(c) }
    mutating func remove(_ c: PositionCredentials)     { self.positionCreds.remove(c) }
    mutating func remove(_ c: WorktimeCredentials)     { self.worktimeCreds.remove(c) }
    mutating func remove(_ c: AppointmentCredentials)  { self.appointmentCreds.remove(c) }
    mutating func remove(_ c: StatisticCredentials)    { self.statisticCreds.remove(c) }
    mutating func remove(_ c: NotificationCredentials) { self.notifyCreds.remove(c) }
    mutating func remove(_ c: SalaryCredentials)       { self.salaryCreds.remove(c) }
    mutating func remove(_ c: FinanceCredentials)      { self.financeCreds.remove(c) }
    
    mutating func contains(_ c: ClientCredentials)       -> Bool { self.clientCreds.contains(c) }
    mutating func contains(_ c: SalonCredentials)        -> Bool { self.salonCreds.contains(c) }
    mutating func contains(_ c: EmployeeCredentials)     -> Bool { self.employeeCreds.contains(c) }
    mutating func contains(_ c: ProcedureCredentials)    -> Bool { self.procedureCreds.contains(c) }
    mutating func contains(_ c: PositionCredentials)     -> Bool { self.positionCreds.contains(c) }
    mutating func contains(_ c: WorktimeCredentials)     -> Bool { self.worktimeCreds.contains(c) }
    mutating func contains(_ c: AppointmentCredentials)  -> Bool { self.appointmentCreds.contains(c) }
    mutating func contains(_ c: StatisticCredentials)    -> Bool { self.statisticCreds.contains(c) }
    mutating func contains(_ c: NotificationCredentials) -> Bool { self.notifyCreds.contains(c) }
    mutating func contains(_ c: SalaryCredentials)       -> Bool { self.salaryCreds.contains(c) }
    mutating func contains(_ c: FinanceCredentials)      -> Bool { self.financeCreds.contains(c) }
}

File: Android/shared/src/main/kotlin/shared/dto/optionSet/CredentialsSet.kt
package shared.dto.optionSet

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
class CredentialsSet {
    var clientCreds      : MutableSet<ClientCredentials> = mutableSetOf()
    var salonCreds       : MutableSet<SalonCredentials> = mutableSetOf()
    var employeeCreds    : MutableSet<EmployeeCredentials> = mutableSetOf()
    var procedureCreds   : MutableSet<ProcedureCredentials> = mutableSetOf()
    var positionCreds    : MutableSet<PositionCredentials> = mutableSetOf()
    var worktimeCreds    : MutableSet<WorktimeCredentials> = mutableSetOf()
    var appointmentCreds : MutableSet<AppointmentCredentials> = mutableSetOf()
    var statisticCreds   : MutableSet<StatisticCredentials> = mutableSetOf()
    var notificationCreds: MutableSet<NotificationCredentials> = mutableSetOf()
    var financeCreds     : MutableSet<FinanceCredentials> = mutableSetOf()
    var salaryCreds      : MutableSet<SalaryCredentials> = mutableSetOf()

    companion object{
        fun all(): CredentialsSet {
            var set = CredentialsSet()
            set.clientCreds = ClientCredentials.entries.toMutableSet()
            set.salonCreds = SalonCredentials.entries.toMutableSet()
            set.employeeCreds = EmployeeCredentials.entries.toMutableSet()
            set.procedureCreds = ProcedureCredentials.entries.toMutableSet()
            set.positionCreds = PositionCredentials.entries.toMutableSet()
            set.worktimeCreds = WorktimeCredentials.entries.toMutableSet()
            set.appointmentCreds = AppointmentCredentials.entries.toMutableSet()
            set.statisticCreds = StatisticCredentials.entries.toMutableSet()
            set.notificationCreds = NotificationCredentials.entries.toMutableSet()
            set.financeCreds = FinanceCredentials.entries.toMutableSet()
            set.salaryCreds = SalaryCredentials.entries.toMutableSet()
            return set
        }

        fun none(): CredentialsSet {
            return CredentialsSet()
        }
    }

    public fun add(c: ClientCredentials)      : Boolean { return this.clientCreds.add(c) }
    public fun add(c: SalonCredentials)       : Boolean { return this.salonCreds.add(c) }
    public fun add(c: EmployeeCredentials)    : Boolean { return this.employeeCreds.add(c) }
    public fun add(c: ProcedureCredentials)   : Boolean { return this.procedureCreds.add(c) }
    public fun add(c: PositionCredentials)    : Boolean { return this.positionCreds.add(c) }
    public fun add(c: WorktimeCredentials)    : Boolean { return this.worktimeCreds.add(c) }
    public fun add(c: AppointmentCredentials) : Boolean { return this.appointmentCreds.add(c) }
    public fun add(c: StatisticCredentials)   : Boolean { return this.statisticCreds.add(c) }
    public fun add(c: NotificationCredentials): Boolean { return this.notificationCreds.add(c) }
    public fun add(c: FinanceCredentials)     : Boolean { return this.financeCreds.add(c) }
    public fun add(c: SalaryCredentials)      : Boolean { return this.salaryCreds.add(c) }

    public fun remove(c: ClientCredentials)      : Boolean { return this.clientCreds.remove(c) }
    public fun remove(c: SalonCredentials)       : Boolean { return this.salonCreds.remove(c) }
    public fun remove(c: EmployeeCredentials)    : Boolean { return this.employeeCreds.remove(c) }
    public fun remove(c: ProcedureCredentials)   : Boolean { return this.procedureCreds.remove(c) }
    public fun remove(c: PositionCredentials)    : Boolean { return this.positionCreds.remove(c) }
    public fun remove(c: WorktimeCredentials)    : Boolean { return this.worktimeCreds.remove(c) }
    public fun remove(c: AppointmentCredentials) : Boolean { return this.appointmentCreds.remove(c) }
    public fun remove(c: StatisticCredentials)   : Boolean { return this.statisticCreds.remove(c) }
    public fun remove(c: NotificationCredentials): Boolean { return this.notificationCreds.remove(c) }
    public fun remove(c: FinanceCredentials)     : Boolean { return this.financeCreds.remove(c) }
    public fun remove(c: SalaryCredentials)      : Boolean { return this.salaryCreds.remove(c) }

    public fun contains(c: ClientCredentials)      : Boolean { return this.clientCreds.contains(c) }
    public fun contains(c: SalonCredentials)       : Boolean { return this.salonCreds.contains(c) }
    public fun contains(c: EmployeeCredentials)    : Boolean { return this.employeeCreds.contains(c) }
    public fun contains(c: ProcedureCredentials)   : Boolean { return this.procedureCreds.contains(c) }
    public fun contains(c: PositionCredentials)    : Boolean { return this.positionCreds.contains(c) }
    public fun contains(c: WorktimeCredentials)    : Boolean { return this.worktimeCreds.contains(c) }
    public fun contains(c: AppointmentCredentials) : Boolean { return this.appointmentCreds.contains(c) }
    public fun contains(c: StatisticCredentials)   : Boolean { return this.statisticCreds.contains(c) }
    public fun contains(c: NotificationCredentials): Boolean { return this.notificationCreds.contains(c) }
    public fun contains(c: FinanceCredentials)     : Boolean { return this.financeCreds.contains(c) }
    public fun contains(c: SalaryCredentials)      : Boolean { return this.salaryCreds.contains(c) }

    enum class ClientCredentials {
        @SerialName("edit")
        EDIT,

        @SerialName("visibleBaseInfo")
        VISIBLE_BASE_INFO,

        @SerialName("visibleContacts")
        VISIBLE_CONTACTS,
    }

    enum class SalonCredentials {
        @SerialName("editBaseInfo")
        EDIT_BASE_INFO,
    }

    enum class EmployeeCredentials {
        @SerialName("edit")
        EDIT,

        @SerialName("visible")
        VISIBLE,
    }

    enum class ProcedureCredentials {
        @SerialName("ownEdit")
        OWN_EDIT,

        @SerialName("otherEdit")
        OTHER_EDIT,
    }

    enum class PositionCredentials {
        @SerialName("edit")
        EDIT,
    }

    enum class WorktimeCredentials {
        @SerialName("ownEdit")
        OWN_EDIT,

        @SerialName("otherEdit")
        OTHER_EDIT,
    }

    enum class AppointmentCredentials {
        @SerialName("all")
        ALL,
    }

    enum class StatisticCredentials {
        @SerialName("all")
        ALL,
    }

    enum class NotificationCredentials {
        @SerialName("all")
        ALL,
    }

    enum class FinanceCredentials {
        @SerialName("all")
        ALL,
    }

    enum class SalaryCredentials {
        @SerialName("all")
        ALL,
    }
}